src/test/java/codeu/controller/AdminServletTest.java

```diff
@@ -0,0 +1,176 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package codeu.controller;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.UUID;
+
+import codeu.model.data.User;
+import codeu.model.data.Conversation;
+import codeu.model.data.Message;
+
+import java.util.List;
+import java.util.ArrayList;
+import codeu.model.store.basic.UserStore;
+import codeu.model.store.basic.ConversationStore;
+import codeu.model.store.basic.MessageStore;
+import javax.servlet.RequestDispatcher;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class AdminServletTest {
+
+  private AdminServlet adminServlet;
+  private HttpServletRequest mockRequest;
+  private HttpServletResponse mockResponse;
+  private RequestDispatcher mockRequestDispatcher;
+  private UserStore mockUserStore;
+  private ConversationStore mockConvoStore;
+  private MessageStore mockMessageStore;
+
+  @Before
+  public void setup() {
+    adminServlet = new AdminServlet();
+    mockRequest = Mockito.mock(HttpServletRequest.class);
+    mockResponse = Mockito.mock(HttpServletResponse.class);
+    mockRequestDispatcher = Mockito.mock(RequestDispatcher.class);
+
+    mockUserStore = Mockito.mock(UserStore.class);
+    mockConvoStore = Mockito.mock(ConversationStore.class);
+    mockMessageStore = Mockito.mock(MessageStore.class);
+
+    User fakeUser =
+        new User(
+            UUID.randomUUID(),
+            "test username",
+            "$2a$10$bBiLUAVmUFK6Iwg5rmpBUOIBW6rIMhU1eKfi3KR60V9UXaYTwPfHy",
+            Instant.now());
+
+    mockUserStore.addUser(fakeUser);
+
+    adminServlet.setUserStore(mockUserStore);
+    adminServlet.setConversationStore(mockConvoStore);
+    adminServlet.setMessageStore(mockMessageStore);
+
+    Mockito.when(mockRequest.getRequestDispatcher("/WEB-INF/view/admin.jsp"))
+        .thenReturn(mockRequestDispatcher);
+  }
+
+  @Test
+  public void testDoGet() throws IOException, ServletException {
+    adminServlet.doGet(mockRequest, mockResponse);
+    mockRequest.getParameter("username");
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+  }
+
+  @Test
+  public void testGetStats() throws IOException, ServletException {
+    Mockito.when(mockRequest.getParameter("username")).thenReturn(null);
```

Comment:
Explain this line, does it mean that when you request the parameter "username", you will get null in return? If yes, why is this needed?

Reply:
Yes thats exactly what it does. This is necessary because of the way I formatted the AdminServlet. Essentially, once a button is pushed (form is submitted) on admin.jsp, the doPost method is called. Right now I have two things going on in admin.jsp once doPost is called: Adding a potential admin (IF the username parameter is not null) and/or updating the site stats. If the username parameter is null, then don't add another admin, but if it is, which is what I am testing here, then don't add an admin, just update the site stats.

Does that kinda make sense?

Reply:
Got it, thanks for explaining.